src/db_iterator.rs

```diff
@@ -82,12 +82,11 @@ pub struct DBRawIteratorWithThreadMode<'a, D: DBAccess> {
 
 impl<'a, D: DBAccess> DBRawIteratorWithThreadMode<'a, D> {
     pub(crate) fn new(db: &D, readopts: ReadOptions) -> Self {
-        unsafe {
-            Self {
-                inner: ffi::rocksdb_create_iterator(db.inner(), readopts.inner),
-                _readopts: readopts,
-                db: PhantomData,
-            }
+        let ptr = unsafe { ffi::rocksdb_create_iterator(db.inner(), readopts.inner) };
+        Self {
+            inner: std::ptr::NonNull::new(ptr).unwrap(),
```

Comment:
Is really this `unwrap` unsafe?

Reply:
As I understand, would panic if the ptr is null here. I like the current version is much better. Because bellow `ffi::rocksdb_iter_valid(self.inner)` function checks for null pointer `self.inner` and if `self.inner` is null then iterator just is invalid? No panic.

Reply:
My thinking here is that rocksdb_create_iterator and rocksdb_create_iterator_cf always return non-null, e.g.:

```c++
rocksdb_iterator_t* rocksdb_create_iterator(
    rocksdb_t* db,
    const rocksdb_readoptions_t* options) {
  rocksdb_iterator_t* result = new rocksdb_iterator_t;
  result->rep = db->rep->NewIterator(options->rep);
  return result;
}
```

and this merely takes advantage of this.  In practice there’s currently no way the unwrap to panic.

Reply:
Potentially, `new` could return a `null` pointer. But in such case, the problem would be in the `rocksdb` library rather than in rust code. Definitely, you should leave some comments about this.

Reply:
> new could return a null pointer.

Ah, I see the code is compiled with `-fno-exceptions`.  So yes, this should be separately fixed in the C wrapper functions and at this point the unwrap would behave like memory allocation error always does in Rust.  By the way:

> Because bellow ffi::rocksdb_iter_valid(self.inner) function checks for null pointer self.inner

This is not true:

```rust
unsigned char rocksdb_iter_valid(const rocksdb_iterator_t* iter) {
  return iter->rep->Valid();
}
```

None of the C wrapper functions check if the pointer is not-null.

Reply:
Ok. You've convinced me 😄

Reply:
Thanks for your investigation 👍🏻